src/Callable.cpp

```diff
@@ -0,0 +1,178 @@
+                   << kind << " of type '" << a.type << "' and dimension " << (int)a.dimensions << ".\n";
+    }
+}
+
+void Callable::check_qcci(size_t argc, const QuickCallCheckInfo *actual_qcci) const {
+    const size_t required_arg_count = contents->quick_call_check_info.size();
+    if (argc == required_arg_count) {
+        const QuickCallCheckInfo *expected_qcci = contents->quick_call_check_info.data();
+        for (size_t i = 0; i < argc; i++) {
+            if (actual_qcci[i] != expected_qcci[i]) {
+                do_check_fail(i, argc, "calling");
+            }
+        }
+    } else {
+        do_check_fail(-1, argc, "calling");
+    }
+}
+
+void Callable::check_fcci(size_t argc, const FullCallCheckInfo *actual_fcci) const {
+    // Lazily create full_call_check_info upon the first call to make_std_function().
+    if (contents->full_call_check_info.empty()) {
+        contents->full_call_check_info.reserve(contents->jit_cache.arguments.size());
+        for (const Argument &a : contents->jit_cache.arguments) {
+            const auto fcci = a.is_scalar() ? Callable::make_scalar_fcci(a.type) : Callable::make_buffer_fcci(a.type, a.dimensions);
+            contents->full_call_check_info.push_back(fcci);
+        }
+    }
+
+    const size_t required_arg_count = contents->full_call_check_info.size();
+    if (argc == required_arg_count) {
+        const FullCallCheckInfo *expected_fcci = contents->full_call_check_info.data();
+        for (size_t i = 0; i < argc; i++) {
+            if (!Callable::is_compatible_fcci(actual_fcci[i], expected_fcci[i])) {
+                do_check_fail(i, argc, "defining");
+            }
+        }
+    } else {
+        do_check_fail(-1, argc, "defining");
+    }
+}
+
+// Entry point used from the std::function<> variant; we can skip the check_qcci() stuff
+// since we verified the signature when we created the std::function, so incorrect types or counts
+// should be impossible.
+/*static*/ int Callable::call_argv_fast(size_t argc, const void *const *argv) const {
+    // Callable should enforce these, so we can use assert() instead of internal_assert()
+    assert(contents->jit_cache.jit_target.has_feature(Target::UserContext));
+    assert(contents->jit_cache.arguments[0].name == "__user_context");
+
+    JITUserContext *context = *(JITUserContext **)const_cast<void *>(argv[0]);
+    assert(context != nullptr);
+
+    JITFuncCallContext jit_call_context(context, contents->saved_jit_handlers);
+
+    // debug(2) << "Calling jitted function\n";
+    int exit_status = contents->jit_cache.call_jit_code(contents->jit_cache.jit_target, argv);
+    // debug(2) << "Back from jitted function. Exit status was " << exit_status << "\n";
+
+    // If we're profiling, report runtimes and reset profiler stats.
+    contents->jit_cache.finish_profiling(context);
+
+    // Don't call jit_call_context.finalize(): if no error handler was installed,
```

Comment:
How does the error message make it out to the user in the case where they don't install a custom error handler?

Reply:
Exactly as we do for AOT calls: we return an error code, which the caller must check.

Reply:
The default for AOT calls is the pipeline calls halide_error, which prints a message then calls abort()

Reply:
It's common for people to override this to log the error string in some other way and then return. The default JIT behavior is that the error string is thrown as an exception when the pipeline returns. This Callable design seems to just drop the error string though and return a non-zero value.